src/arith/transitive_comparison_analyzer.cc

```diff
@@ -0,0 +1,683 @@
+
+  // Utility function to add a new known statement
+  auto declare_known = [&](Comparison cmp) {
+    auto& prev_knowns = compared_to_x[cmp.rhs_];
+
+    for (auto& prev_known : prev_knowns) {
+      if (prev_known.Implies(cmp)) {
+        return;
+      }
+    }
+
+    if (cmp.rhs_ != rhs_key && !seen.count(cmp.rhs_)) {
+      to_visit.insert(cmp.rhs_);
+      seen.insert(cmp.rhs_);
+    }
+
+    for (auto& prev_known : prev_knowns) {
+      if (cmp.Implies(prev_known)) {
+        prev_known = cmp;
+        return;
+      }
+    }
+
+    prev_knowns.push_back(cmp);
+  };
+
+  // Initialize the search based on any known (in)equalities that use
+  // the LHS of the comparison.
+  for (const auto& known : knowns_) {
+    if (auto normalized = known.WithLHS(lhs_key)) {
+      declare_known(normalized.value());
+    }
+  }
+  for (const auto& known : scoped_knowns_) {
+    if (auto normalized = known.WithLHS(lhs_key)) {
+      declare_known(normalized.value());
+    }
+  }
+
+  // Walk through the space of all comparisons that can be made with
+  // LHS.
+  while (to_visit.size()) {
+    Key middle_key = *to_visit.begin();
+    to_visit.erase(to_visit.begin());
+
+    std::vector<Comparison>& prev_knowns_using_middle = compared_to_x.at(middle_key);
+    ICHECK(compared_to_x.count(middle_key));
+
+    std::vector<Comparison> new_knowns_using_lhs;
+
+    auto attempt_transitive = [&](Comparison cmp) {
+      ICHECK(cmp.IsNormalized());
+
+      Key right_key = cmp.rhs_;
+
+      if (right_key == lhs_key) {
+        return;
+      }
+
+      for (const auto& prev : prev_knowns_using_middle) {
+        CompareResult new_result = CompareResult::kUnknown;
+        int64_t new_offset = prev.offset_ + cmp.offset_;
+
+        if (prev.result_ == CompareResult::kEQ) {
+          // x == y + c1 && y OP z + c2, x OP z + (c1 + c2)
+          new_result = cmp.result_;
+        } else if (cmp.result_ == CompareResult::kEQ) {
+          // x OP y + c1 && y == z + c2, x OP z + (c1 + c2)
+          new_result = prev.result_;
+        } else if (prev.result_ == cmp.result_ &&
+                   (prev.result_ == CompareResult::kLE || prev.result_ == CompareResult::kGE)) {
+          // x <= y + c1 && y <= z + c2, x <= z + (c1 + c2)
+          // x >= y + c1 && y >= z + c2, x >= z + (c1 + c2)
+          //
+          // This condition is much simpler to write than the
+          // equivalent handling of < or of >, which is why the
+          // inequalities are normalized to <= and to >=.
+          new_result = prev.result_;
+        }
+
+        if (new_result != CompareResult::kUnknown) {
+          Comparison new_known(lhs_key, right_key, new_offset, new_result);
+          new_knowns_using_lhs.push_back(new_known);
+        }
+      }
+    };
+
+    // Attempt to prove a new comparison using one of the original
+    // known comparisons.  We want to find a known such that
+    // `(LHS OP1 middle) && (middle OP2 right)` can be simplified
```

Comment:
nit: Switch of convention from LHS to right instead of RHS. Maybe just use left/middle/right

Reply:
I'm not sure the best way to phrase this part.  The switch of convention is intentional, as the `right` in this comment might not be the RHS of the desired comparison.  However, this new comparison will always use the LHS of the   For example, if it is known that `a <= b`, `b <= c`, and `c <= d`, then the first step is to prove that `a <= c`.  While proving that, `a` is LHS, `b` is middle, and `c` is right.

I've updated the comment for clarity, with an example.